src/viper/prep.ml

```diff
@@ -0,0 +1,145 @@
+open Common
+open Source
+open Mo_def.Syntax
+open Traversals
+
+module T = Mo_types.Type
+module Cons = Mo_types.Cons
+module Arrange = Mo_def.Arrange
+
+(*
+  Monomorphisation goal. For example, a function call f<Int,Bool>(a,b,c)
+  yields a mono_goal of {id="f", typs=[Int,Bool]}
+*)
+type mono_goal = { mg_id : string; mg_typs : T.typ list; }
+
+let compare_mono_goal (g1 : mono_goal) (g2 : mono_goal) =
+  match String.compare g1.mg_id g2.mg_id with
+  | 0 -> List.compare T.Ord.compare g1.mg_typs g2.mg_typs
+  | ord -> ord
+
+type dec_field_template = { dft_id : string; dft_mk : T.typ list -> dec_field }
+
+let string_of_mono_goal (g : mono_goal) : string =
+  String.concat "$" (g.mg_id :: List.map (fun t ->
+    match T.normalize t with
+    | T.Prim T.Int  -> "Int"
+    | T.Prim T.Nat  -> "Nat"
+    | T.Prim T.Bool -> "Bool"
+    | _ -> unsupported Source.no_region (Mo_types.Arrange_type.typ t)) g.mg_typs)
+
+let mono_calls_visitor (stk : mono_goal Stack.t) : visitor =
+  { visit_exp = (function
+    | {it = CallE({it = VarE v; at = v_at; note = v_note},inst,e); at; note} ->
```

Comment:
Curious: how would you handle first-class uses of a polymorphic functions? Create a record of monomorphic ones?

Reply:
Yes, in case of a higher-rank usage that would be a plausible strategy. A record of monomorphic functions could indeed represent a finite set of instantiations of a polymorphic function.

However, a first-class use is not necessarily higher-rank. E.g. in `map(f, xs)` it wouldn't be necessary to turn `f` into a record of functions because all elements are of a single type. So we will end up with e.g. `map$Int(f$Int, xs)` if the elements are integers.

Actual higher-rank usage is rare in practice and I'm not sure if we should invest effort into supporting it.